src/metabase/notification/send.clj

```diff
@@ -181,23 +181,100 @@
+
+(defn- subscription->deadline
+  "Calculates a deadline for notification execution with priority based on frequency.
+
+  More frequent notifications (based on cron schedule) receive shorter deadlines,
+  which results in higher priority when multiple notifications are scheduled at
+  the same time."
+  [{:keys [type cron_schedule]}]
+  (let [deadline-bonus (case type
+                         :notification-subscription/cron
+                         (let [avg-interval-seconds (avg-interval-seconds cron_schedule 10)
+                               interval-less-than   (fn [duration] (< avg-interval-seconds (t/as duration :seconds)))]
+                           (cond
+                             (interval-less-than (t/minutes 1))  (t/seconds 5)
+                             (interval-less-than (t/minutes 5))  (t/seconds 10)
+                             (interval-less-than (t/minutes 30)) (t/seconds 15)
+                             (interval-less-than (t/hours 1))    (t/seconds 30)
+                             :else                               (t/seconds 60)))
+                         ;; default to 30 seconds for non-cron subscriptions
+                         (t/seconds 30))]
+    (t/plus (t/local-date-time) deadline-bonus)))
+
+;; A notification that can be put into a queue but has equal checks based on its ID
+(deftype NotificationQueueEntry [id deadline]
+  Object
+  (equals [this  other]
+    (and (instance? NotificationQueueEntry other)
+         (= (.id this) (.id ^NotificationQueueEntry other))))
+
+  (hashCode [this]
+    (hash (.id this)))
+  Comparable
+  (compareTo [this other]
+    (compare (.id this) (.id ^NotificationQueueEntry other))))
+
+(defn- deadline-comparator
+  "Comparator for sorting notifications by deadline."
+  [a b]
+  (compare (.deadline ^NotificationQueueEntry a) (.deadline ^NotificationQueueEntry b)))
+
 (defprotocol NotificationQueueProtocol
   "Protocol for notification queue implementations."
   (put-notification!  [this notification] "Add a notification to the queue. If a notification with the same id is already in the queue, replace it.")
   (take-notification! [this]              "Take the next notification from the queue, blocking if none available."))
 
 (deftype ^:private NotificationQueue
-         [^java.util.LinkedList                     ids-list
+         [^java.util.PriorityQueue                  items-list
           ^java.util.concurrent.ConcurrentHashMap   id->notification
           ^java.util.concurrent.locks.ReentrantLock queue-lock
           ^java.util.concurrent.locks.Condition     not-empty-cond]
   NotificationQueueProtocol
   (put-notification! [_ notification]
-    (let [id (or (:id notification) (str (random-uuid)))]
-      (.lock queue-lock)
+    (let [id   (or (:id notification) (str (random-uuid)))
+          item (NotificationQueueEntry. id (subscription->deadline (:triggering_subscription notification)))]
       (try
+        (.lock queue-lock)
```

Comment:
Why did you move the locking inside the try? I'm slightly nervous that an exception, say timeout, will cause this to release the lock even if we were not the one holding it.

Reply:
hmm, that's true. I'll move it out

Reply:
that's so subtle. thanks for flagging.